sip/sip-002-smart-contract-language.md

```diff
@@ -0,0 +1,493 @@
+6. Defining of constants and functions are allowed for simplifying
+   code. However, these are purely syntactic. If a definition cannot be
+   inlined, the contract will be rejected as illegal.
+7. Transactions are specified via `define-tx` statement with function
+   names. Arguments to the function must specify their types.
+
+Transactions return a boolean result. If the transaction returns
+`true`, then it is considered valid, and any changes made to the
+blockchain state will be materialized. If a transaction returns
+`false`, the transaction will be considered invalid, and the
+transaction will have _no effect_ on the smart contract's state,
+except for a transaction fee debit (in the case of on-chain
+transactions).
+
+## Inter-Contract Calls
+
+A smart contract may call functions from other smart contracts using a
+`(contract-call)` function. This function accepts a function name and
+the smart contract's _identifier_ as input.  A smart contract's
+identifier is a hash of the smart contract's definition, represented
+as a Stacks address with a specific "smart contract" version byte. The
+smart contract identifier is a _principal_.
+
+For example, to call the function `register-name` in a smart contract,
+you would use:
+
+```scheme
+(contract-call
+    'SC3H92H297DX3YDPFHZGH90G8Z4NPH4VE8E83YWAQ
+    'register-name
+    name-to-register)
+```
+
+This function returns a boolean-- the return value of the called smart
+contract function. Note that if a called smart contract returns
+`false`, it is guaranteed to not alter any smart contract state
+whatsoever. Of course, any transaction fees paid for the execution
+of that function will not be returned.
+
+The following limitations are imposed on contract calls:
+
+1. No dynamic dispatch. At the time of the smart contract creation,
+   any contracts being called must be specified. Future designs may
+   enable this by allowing contract principals to be supplied as
+   transaction arguments, however, on initial release, we believe
+   dynamic invocation to be too dangerous to support.
+2. Called smart contracts _must_ exist at the time of creation.
+3. No cycles may exist in the call graph of a smart contract. This
+   prevents recursion (and re-entrancy bugs). Such structures can
+   be detected with static analysis of the call graph, and will be
+   rejected by the network.
+
+A key benefit of the static analyzability of this smart contracting
+language is that _all_ transactions that can possibly be called from
+the outermost transaction can be known _a priori_ so that a user can
+be warned about all side effects before signing a transaction.
+
+## Principals and Transaction Verification
+
+The language provides a primitive for checking whether or not the
+smart contract transaction was signed by a particular
+_principal_. Principals are a specific type in the smart contracting
+language which represent a signing entity (roughly equivalent to a
```

Comment:
Not a signing entity, but a spending entity?  A smart contract does not sign something, but both an incoming transaction and a smart contract method routine can spend.

Reply:
Hmm, I'm okay with spending entity as a placeholder for now, but we should come up with a better terminology at some point

Reply:
Updated to spending entity in the SIP